### PR TITLE
Reduce allocations for `ServiceTalkBufferAllocator.wrap(byte[], int, int)`

### PR DESCRIPTION
**Pull request by idelpivnitskiy**
Motivation:

Default impl in `BufferAllocator` takes a slice of a wrapped original byte-array. We can avoid double wrapping with `NettyBuffer` if we implement this method in `ServiceTalkBufferAllocator`.

Modifications:

- Override `BufferAllocator.wrap(byte[], int, int)` to wrap with `NettyBuffer` only one time;
- Enhance `BufferAllocatorsTest` to validate wrapping and ensure the wrapped `Buffer` always has readerIndex set to 0;

Result:

Minus 2 object allocations when users wrap a byte-array partially.

### CODE REVIEW

File: servicetalk-buffer-netty/src/test/java/io/servicetalk/buffer/netty/BufferAllocatorsTest.java [hunk 1]
```diff
@@ -76,6 +84,42 @@ void testReadOnlyDirectBuffer(BufferAllocator allocator) {
         assertBuffer(allocator.wrap(ByteBuffer.allocateDirect(16).asReadOnlyBuffer()), true);
     }
```
Comment (daschl) [inline servicetalk-buffer-netty/src/test/java/io/servicetalk/buffer/netty/BufferAllocatorsTest.java:86]: consider also testing negative case - i.e length is negative and/or offset is more than bytes length.

**Review by daschl — COMMENTED** [review]
minor comment about test coverage, optimization looks good!

File: servicetalk-buffer-netty/src/main/java/io/servicetalk/buffer/netty/ServiceTalkBufferAllocator.java [hunk 2]
```diff
@@ -159,6 +159,17 @@ public Buffer wrap(byte[] bytes) {
...
     }
 
+    @Override
+    public Buffer wrap(final byte[] bytes, final int offset, final int len) {
+        if (len == 0) {
+            return EMPTY_BUFFER;
+        }
+        final Buffer b = wrap(bytes);
+        b.readerIndex(offset);
+        b.writerIndex(offset + len);
```
Comment (bryce-anderson) [inline servicetalk-buffer-netty/src/main/java/io/servicetalk/buffer/netty/ServiceTalkBufferAllocator.java:169]: In my minds eye this might be an unexpected behavior change: the resulting buffer will have different reader and writer indexes than expected as if they had sliced the `byte[]` themselves beforehand. This could be a trip hazard as users often will reset the reader index to 0 to reuse the buffer efficiently.

**Review by daschl — APPROVED** [review]

**Review by bryce-anderson — APPROVED** [review]
Looks good after fixing style check. I think you can still eliminate a `NettyBuffer` allocation if you override the `wrap(byte[], int, int)` method to slice the netty buffer before wrapping with `NettyBuffer`, although it may be that it was stack allocated all along.

File: servicetalk-buffer-netty/src/test/java/io/servicetalk/buffer/netty/BufferAllocatorsTest.java (hunk 1 above)
Comment (idelpivnitskiy) [reply servicetalk-buffer-netty/src/test/java/io/servicetalk/buffer/netty/BufferAllocatorsTest.java:113]: Added `testWrapPartialByteArrayWithIllegalArguments`

**Review by idelpivnitskiy — COMMENTED** [review]
Forgot to publish my reply comments y'day

**Review by bryce-anderson — APPROVED** [review]
